zkfacade/src/main/java/com/yahoo/vespa/curator/ThreadLockInfo.java

```diff
@@ -0,0 +1,125 @@
+
+    private static final int MAX_COMPLETED_LOCK_INFOS_SIZE = 10;
+    private static final ConcurrentLinkedDeque<LockInfo> completedLockInfos = new ConcurrentLinkedDeque<>();
+
+    private static final AtomicInteger invokeAcquireCount = new AtomicInteger(0);
+    private static final AtomicInteger inCriticalRegionCount = new AtomicInteger(0);
+    private static final AtomicInteger acquireTimedOutCount = new AtomicInteger(0);
+    private static final AtomicInteger lockAcquiredCount = new AtomicInteger(0);
+    private static final AtomicInteger locksReleasedCount = new AtomicInteger(0);
+
+    private static final AtomicInteger failedToAcquireReentrantLockCount = new AtomicInteger(0);
+    private static final AtomicInteger noLocksErrorCount = new AtomicInteger(0);
+    private static final AtomicInteger timeoutOnReentrancyErrorCount = new AtomicInteger(0);
+
+    private final Thread thread;
+    private final String lockPath;
+    private final ReentrantLock lock;
+
+    /** The locks are reentrant so there may be more than 1 lock for this thread. */
+    private final ConcurrentLinkedQueue<LockInfo> lockInfos = new ConcurrentLinkedQueue<>();
+
+    public static int invokeAcquireCount() { return invokeAcquireCount.get(); }
+    public static int inCriticalRegionCount() { return inCriticalRegionCount.get(); }
+    public static int acquireTimedOutCount() { return acquireTimedOutCount.get(); }
+    public static int lockAcquiredCount() { return lockAcquiredCount.get(); }
+    public static int locksReleasedCount() { return locksReleasedCount.get(); }
+    public static int noLocksErrorCount() { return noLocksErrorCount.get(); }
+    public static int failedToAcquireReentrantLockCount() { return failedToAcquireReentrantLockCount.get(); }
+    public static int timeoutOnReentrancyErrorCount() { return timeoutOnReentrancyErrorCount.get(); }
+    public static List<ThreadLockInfo> getThreadLockInfos() { return List.copyOf(locks.values()); }
+
+    /** Returns the per-thread singleton ThreadLockInfo. */
+    static ThreadLockInfo getCurrentThreadLockInfo(String lockPath, ReentrantLock lock) {
+        return locks.computeIfAbsent(
+                Thread.currentThread(),
+                currentThread -> new ThreadLockInfo(currentThread, lockPath, lock));
+    }
+
+    ThreadLockInfo(Thread currentThread, String lockPath, ReentrantLock lock) {
+        this.thread = currentThread;
+        this.lockPath = lockPath;
+        this.lock = lock;
+    }
+
+    public String getThreadName() { return thread.getName(); }
+    public String getLockPath() { return lockPath; }
+    public List<LockInfo> getLockInfos() { return List.copyOf(lockInfos); }
+
+    /** Mutable method (see class doc) */
+    void invokingAcquire(Duration timeout) {
+        invokeAcquireCount.incrementAndGet();
+        inCriticalRegionCount.incrementAndGet();
+        lockInfos.add(LockInfo.invokingAcquire(lock.getHoldCount(), timeout));
```

Comment:
Is this done to track the re-entry count?

Reply:
The getHoldCount() ? Yes - just to sanity-check it